### PR TITLE
[SPARK-37216][SQL] Add the Hive macro functionality to SparkSQL

### PR DESCRIPTION
**Pull request by hgs19921112**
### What changes were proposed in this pull request?

Add the Hive macro functionality to SparkSQL

### Why are the changes needed?

Some Hive sql can move to SparkSQL Smoothly

### Does this PR introduce _any_ user-facing change?

Some new DDL like 'create temparory macro ...'

### How was this patch tested?

unit test

Authored-by: hgs19921112 <haoguangshi@gmail.com>

### CODE REVIEW

**Comment by hgs19921112** [comment]
Can anayone have a see? If this pr is useless , I will close it.

**Comment by xkrogen** [comment]
This seems to have some overlap with the recent functionality announced by Databricks (as an internal/proprietary feature): https://databricks.com/blog/2021/10/20/introducing-sql-user-defined-functions.html
Is that understanding correct? Can you compare?
I am wondering if, similar to the Databricks feature, OSS Spark should have a more holistic way of handling this as opposed to copying Hive's syntax for a limited subset of the functionality.

cc @mridulm @wmoustafa @shardulm94

**Comment by hgs19921112** [comment]
> This seems to have some overlap with the recent functionality announced by Databricks (as an internal/proprietary feature): https://databricks.com/blog/2021/10/20/introducing-sql-user-defined-functions.html Is that understanding correct? Can you compare? I am wondering if, similar to the Databricks feature, OSS Spark should have a more holistic way of handling this as opposed to copying Hive's syntax for a limited subset of the functionality.
> 
> cc @mridulm @wmoustafa @shardulm94
You are right . The functionality in Databricks's Spark is more powerful than this.I did not know about that before you mentioned it.Mybe this is excess.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 1]
```diff
@@ -252,6 +252,8 @@ statement
         '(' columns=multipartIdentifierPropertyList ')'
         (OPTIONS options=propertyList)?                                #createIndex
     | DROP INDEX (IF EXISTS)? identifier ON TABLE? multipartIdentifier #dropIndex
+    | CREATE TEMPORARY MACRO IDENTIFIER '(' colTypeList ')'
```
Comment (ever4Kenny) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:255]: should be:
"| CREATE TEMPORARY MACRO IDENTIFIER '(' colTypeList? ')'"
to support no arg macro.
The colTypeList is optional.

**Pull request closed without merging**